File: src/pulseDetect.cc
#include <thread>
#include "../include/HW.h"
static int numberOfInterrupts = 0;
static void interruptHandler() {
  numberOfInterrupts++;
};
int main(int argc, char ** argv) {
  /*
   This program demonstrates the use of wiringPi to monitor a pin and count the pulses that occur on that pin.
   The pin, physical pin 40, is attached to a hardware based PWM at physical pin 12.  This program sets the
   frequency of the PWM to 1.14 Hz (period 877 ms) and the duty cycle to 50%.  Half the time the pulse will be
   high, half low.
  
   Two types of pulse detectors are demonstrated.  The first is interrupt driven.  An interrupt is attached to
   the rising edge of the signal on physical pin 40.  That handler simply increments a counter.  The second example
   polls on a background thread every 25 msec.  When the signal is high for a sufficient amount of time, the pulse
   count in incremented.  Note that this thread is used to terminate the program by either of two means - either 100
   pulses are detected, or a watchdog timer expires.  This means that the print statement should always say that the
   polling routine's count should be 100 unless the watchdog causes the termination.

   Copyright (c) 2021
   Mark Broihier
  */
  HW hardwareInterface;
  int numberOfPulses = 0;
  int highestHighCount = 0;
  int highestLowCount = 0;
  auto pulseDetector2 = [&numberOfPulses, &hardwareInterface, &highestHighCount, &highestLowCount]() {
                          int const THRESHOLD = 6;
                          int const EXIT_WHEN_VALID = 100;
                          int const EXIT_WHEN_INVALID = 1000;
                          int levelHighCount = 0;
                          int levelLowCount = 0;
                          int watchdog = 0;
                          struct timeval tv = {0, 25000}; // check every 25 milliseconds
                          while (numberOfPulses < EXIT_WHEN_VALID) {
                            select(0, NULL, NULL, NULL, &tv);  // wait and give up control
                            tv = {0, 25000}; // reset timeout                        
                            if (hardwareInterface.signalIn() == 1) {
                              levelLowCount = 0;
                              levelHighCount++;
                              if (levelHighCount > highestHighCount) {
                                highestHighCount = levelHighCount;
                              }
                            } else {
                              levelHighCount = 0;
                              levelLowCount++;
                              if (levelLowCount > highestLowCount) {
                                highestLowCount = levelLowCount;
                              }
                            }
                            if (levelHighCount == THRESHOLD) {
                              numberOfPulses++;
                              watchdog = 0;
                            }
                            if (watchdog++ > EXIT_WHEN_INVALID) return;  // if physical pins 12 and 40 are not connected
                                                                         // this stops the thread
                          }
                          hardwareInterface.turnOffPWMPin();
                          return;
                        };
  std::thread background(pulseDetector2);
  hardwareInterface.setInterruptHandler(interruptHandler);
  fprintf(stdout, "Setup PWM for 1.14 Hz 50%% duty cycle\n");
  hardwareInterface.setupPWMPin(4095, 4096, 0.50);  // clock (divisor), range, duty cycle
  background.join();  // wait for background thread to terminate
  fprintf(stdout, "The interrupt handler observed %d zero to one transitions/pulses\n", numberOfInterrupts);
  fprintf(stdout, "The polling routine observed %d pulses\n", numberOfPulses);
  fprintf(stdout, "Saw as many as %d counts while the level was high and %d counts while the level was low\n",
          highestHighCount, highestLowCount);
  return 0;
}

File: pulseDetect.py
#!/usr/bin/python3
import os
import select
import sys
import threading
import wiringpi

class HW():
    '''
    Class that creates a hardware interface to demonstate pulse detectors
    '''
    def __init__(self):
        '''
        Initialize wiringPi and objects needed
        '''
        if os.getuid() == 0:
            wiringpi.wiringPiSetupGpio()
            self.PWMOutputPin = 18
            self.SignalInputPin = 21
            wiringpi.pinMode(self.PWMOutputPin, wiringpi.PWM_OUTPUT)
            wiringpi.pinMode(self.SignalInputPin, wiringpi.INPUT)
        else:
            print("This object, build with the HW class, requires root privileges")
            sys.exit(-1)

    def signalIn(self):
        '''
        signalIn - read the current signal level needed by the pulse detectors
        '''
        return (wiringpi.digitalRead(self.SignalInputPin))

    def setInterruptHandler(self, interruptHandler ):
        '''
        setInterruptHandler - connect a handler to the signal input
        '''
        wiringpi.wiringPiISR(self.SignalInputPin, wiringpi.INT_EDGE_RISING, interruptHandler)

    def turnOffPWMPin(self):
        '''
        turnOffPWMPin - zero the PWM setting to disable pulses
        '''
        wiringpi.pwmWrite(self.PWMOutputPin, 0)

    def setupPWMPin(self, clock, range_, dutyCycle):
        '''
        setupPWMPin - setup the hardware for pulse generation and start the pulses
        '''
        wiringpi.pwmSetMode(wiringpi.PWM_MODE_MS)
        wiringpi.pwmSetClock(clock)
        wiringpi.pwmSetRange(range_)
        setting = int(range_ * dutyCycle)
        print("Settings for PWM0 are {} clock divisor, {} range, and {} setting".format(clock, range_, setting))
        wiringpi.pwmWrite(self.PWMOutputPin, setting)

class pulseDetector2(threading.Thread):
    def __init__(self, hardwareInterface, sharedVariables):
        '''
        init - constructor - sets up shared objects
        '''
        super(pulseDetector2, self).__init__()
        self.hardwareInterface = hardwareInterface
        self.sharedVariables = sharedVariables

    def run(self):
        '''
        run - thread that polls the input signal in background to count pulses
        '''
        THRESHOLD = 6
        EXIT_WHEN_VALID = 100
        EXIT_WHEN_INVALID = 1000
        levelHighCount = 0
        levelLowCount = 0
        watchdog = 0
        while self.sharedVariables['numberOfPulses'] < EXIT_WHEN_VALID:
            select.select([], [], [], 0.025)
            if self.hardwareInterface.signalIn() == 1:
                levelLowCount = 0
                levelHighCount += 1
                if levelHighCount > self.sharedVariables['highestHighCount']:
                    self.sharedVariables['highestHighCount'] = levelHighCount
            else:
                levelHighCount = 0
                levelLowCount += 1
                if levelLowCount > self.sharedVariables['highestLowCount']:
                    self.sharedVariables['highestLowCount'] = levelLowCount
            if levelHighCount == THRESHOLD:
                self.sharedVariables['numberOfPulses'] += 1
                watchdog = 0
            watchdog += 1
            if watchdog > EXIT_WHEN_INVALID:
                break
        self.hardwareInterface.turnOffPWMPin()

numberOfInterrupts = 0

def interruptHandler():
    global numberOfInterrupts
    numberOfInterrupts += 1
    
def main():
    '''
    Main program for pulseDetect
    '''
    global numberOfInterrupts
    hardwareInterface = HW()
    sharedVariables = {'numberOfPulses' : 0, 'highestHighCount' : 0, 'highestLowCount' : 0 }
    background = pulseDetector2(hardwareInterface, sharedVariables)
    background.start()
    hardwareInterface.setInterruptHandler(interruptHandler)
    print("Setup PWM for 1.14 Hz 50% duty cycle")
    hardwareInterface.setupPWMPin(4095, 4096, 0.50)
    background.join()
    print("The interrup handler observed {} zero to one transitions/pulses".format(numberOfInterrupts))
    print("The polling routine observed {} pulses".format(sharedVariables['numberOfPulses']))
    print("The polling routine saw as many as {} counts while the level was high and {} counts while the level was low".
          format(sharedVariables['highestHighCount'], sharedVariables['highestLowCount']))
if __name__ == '__main__':
    main()
